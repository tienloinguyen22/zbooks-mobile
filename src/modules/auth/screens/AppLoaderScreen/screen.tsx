import React, { useEffect } from 'react';
import { ScreenProps } from '@app/core';
import i18next from 'i18next';
import SplashScreen from 'react-native-splash-screen';
import { navigationService } from '@app/services';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import _ from 'lodash';

type Props = ScreenProps;

const APP_SETTINGS_AND_CURRENT_USER = gql`
  query GetAppSettingsAndCurrentUser {
    appSettings @client {
      language
      primaryColorCode
    }
    currentUser @client {
      isLoggedIn
    }
  }
`;

export const Screen = (_props: Props): JSX.Element => {
  const { data, loading } = useQuery(APP_SETTINGS_AND_CURRENT_USER);
  const language = _.get(data, 'appSettings.language');
  const isLoggedIn = _.get(data, 'currentUser.isLoggedIn');

  useEffect(() => {
    if (loading) {
      return;
    }
    i18next.changeLanguage(language);

    (async (): Promise<void> => {
      if (isLoggedIn) {
        navigationService.setRootHome();
      } else {
        navigationService.setRootLogin();
      }

      SplashScreen.hide();
    })();
  }, [isLoggedIn, language, loading]);

  return <></>;
};
