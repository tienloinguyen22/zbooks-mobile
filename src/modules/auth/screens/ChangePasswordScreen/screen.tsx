import React, { useState } from 'react';
import { Button, Text, Field, Container } from '@app/components';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import { Formik } from 'formik';
import { ScreenProps, showNotification, i18n, handleError } from '@app/core';
import { navigationService, authService } from '@app/services';
import { styles } from './styles';
import { mapStateToProps } from './map_state_to_props';
import { mapDispatchToProps } from './map_dispatch_to_props';

type Props = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps> & ScreenProps;

interface FormData {
  password: string;
  confirmPassword: string;
}

export const Screen = ({ componentId, language }: Props): JSX.Element => {
  const { t } = useTranslation();
  const [isBusy, setIsBusy] = useState<boolean>(false);

  const initialValues: FormData = {
    password: '',
    confirmPassword: '',
  };
  const fieldNames = {
    password: 'password',
    confirmPassword: 'confirmPassword',
  };
  const validationSchema = Yup.object().shape({
    [fieldNames.password]: Yup.string()
      .matches(
        /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,20}$/,
        t('error.invalid', {
          field: t('changePasswordScreen.password'),
        }),
      )
      .required(
        t('error.required', {
          field: t('changePasswordScreen.password'),
        }),
      ),
    [fieldNames.confirmPassword]: Yup.string()
      .oneOf([Yup.ref('password')], t('changePasswordScreen.confirmPasswordDoesNotMatch'))
      .required(
        t('error.required', {
          field: t('changePasswordScreen.confirmPassword'),
        }),
      ),
  });

  const onSubmit = async (values: FormData): Promise<void> => {
    try {
      setIsBusy(true);
      await authService.changePassword(values.password);
      showNotification({
        type: 'SUCCESS',
        message: t('changePasswordScreen.passwordChanged'),
      });
      navigationService.goBack({
        componentId,
      });
    } catch (error) {
      handleError(error, {
        'auth/requires-recent-login': t('changePasswordScreen.requireRecentLogin'),
      });
    } finally {
      setIsBusy(false);
    }
  };

  return (
    <Container
      showHeader
      showBackButton
      componentId={componentId}
      headerTitle={t('changePasswordScreen.changePassword')}
    >
      <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={onSubmit}>
        {(props) => {
          const { values, touched, errors, handleChange, handleBlur, handleSubmit } = props;
          return (
            <>
              <Field
                label={t('changePasswordScreen.password')}
                value={values.password}
                onChangeText={handleChange(fieldNames.password)}
                onBlur={handleBlur(fieldNames.password)}
                error={touched.password && !!errors.password}
                success={touched.password && !errors.password}
                errorMessage={errors.password}
                secureTextEntry
                hasTooltip
                tooltipText={t('common.passwordInfo')}
                tooltipHeight={language === i18n.LANGUAGE_EN ? 140 : 100}
              />
              <Field
                label={t('changePasswordScreen.confirmPassword')}
                value={values.confirmPassword}
                onChangeText={handleChange(fieldNames.confirmPassword)}
                onBlur={handleBlur(fieldNames.confirmPassword)}
                error={touched.confirmPassword && !!errors.confirmPassword}
                success={touched.confirmPassword && !errors.confirmPassword}
                errorMessage={errors.confirmPassword}
                secureTextEntry
              />
              <Button onPress={handleSubmit} disabled={isBusy} style={styles.button}>
                <Text white>{t('changePasswordScreen.changePassword')}</Text>
              </Button>
            </>
          );
        }}
      </Formik>
    </Container>
  );
};
