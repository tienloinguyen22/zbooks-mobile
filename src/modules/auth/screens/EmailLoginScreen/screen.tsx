import React, { useState } from 'react';
import { Button, Text, Field, Container } from '@app/components';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import { Formik } from 'formik';
import { ScreenProps, i18n, screenNames, handleError } from '@app/core';
import { navigationService, authService } from '@app/services';
import { styles } from './styles';
import { mapStateToProps } from './map_state_to_props';
import { mapDispatchToProps } from './map_dispatch_to_props';

type Props = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps> & ScreenProps;

interface FormData {
  email: string;
  password: string;
}

export const Screen = ({ componentId, login, language }: Props): JSX.Element => {
  const { t } = useTranslation();
  const [isBusy, setIsBusy] = useState<boolean>(false);

  const initialValues: FormData = {
    email: '',
    password: '',
  };
  const fieldNames = {
    email: 'email',
    password: 'password',
  };
  const validationSchema = Yup.object().shape({
    [fieldNames.email]: Yup.string()
      .email(
        t('error.invalid', {
          field: t('emailLoginScreen.email'),
        }),
      )
      .required(
        t('error.required', {
          field: t('emailLoginScreen.email'),
        }),
      ),
    [fieldNames.password]: Yup.string()
      .matches(
        /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,20}$/,
        t('error.invalid', {
          field: t('emailLoginScreen.password'),
        }),
      )
      .required(
        t('error.required', {
          field: t('emailLoginScreen.password'),
        }),
      ),
  });

  const onSubmit = async (values: FormData): Promise<void> => {
    try {
      setIsBusy(true);
      const user = await authService.signInWithEmailAndPassword(values.email, values.password);
      login(user);
      if (user.emailVerified) {
        navigationService.setRootHome();
      } else {
        navigationService.setRootEmailVerification();
      }
    } catch (error) {
      handleError(error, {
        'auth/invalid-email': t('emailLoginScreen.wrongEmailOrPassword'),
        'auth/user-disabled': t('emailLoginScreen.userDisabled'),
        'auth/user-not-found': t('emailLoginScreen.wrongEmailOrPassword'),
        'auth/wrong-password': t('emailLoginScreen.wrongEmailOrPassword'),
      });
    } finally {
      setIsBusy(false);
    }
  };

  const forgotPassword = (): void =>
    navigationService.navigateTo({
      componentId,
      screenName: screenNames.ForgotPasswordScreen,
    });

  return (
    <Container showHeader showBackButton componentId={componentId} headerTitle={t('emailLoginScreen.login')}>
      <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={onSubmit}>
        {(props) => {
          const { values, touched, errors, handleChange, handleBlur, handleSubmit } = props;
          return (
            <>
              <Field
                label={t('emailLoginScreen.email')}
                value={values.email}
                onChangeText={handleChange(fieldNames.email)}
                onBlur={handleBlur(fieldNames.email)}
                error={touched.email && !!errors.email}
                success={touched.email && !errors.email}
                errorMessage={errors.email}
              />
              <Field
                label={t('emailLoginScreen.password')}
                value={values.password}
                onChangeText={handleChange(fieldNames.password)}
                onBlur={handleBlur(fieldNames.password)}
                error={touched.password && !!errors.password}
                success={touched.password && !errors.password}
                errorMessage={errors.password}
                secureTextEntry
                hasTooltip
                tooltipText={t('common.passwordInfo')}
                tooltipHeight={language === i18n.LANGUAGE_EN ? 140 : 100}
              />
              <Button onPress={handleSubmit} disabled={isBusy} style={[styles.button, styles.firstButton]}>
                <Text white>{t('emailLoginScreen.login')}</Text>
              </Button>
              <Button onPress={forgotPassword} disabled={isBusy} style={styles.button}>
                <Text white>{t('emailLoginScreen.forgotPassword')}</Text>
              </Button>
            </>
          );
        }}
      </Formik>
    </Container>
  );
};
