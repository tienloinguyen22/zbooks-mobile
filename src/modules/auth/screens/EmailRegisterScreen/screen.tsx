import React, { useState } from 'react';
import { Button, Text, Field, Container } from '@app/components';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import * as _ from 'lodash';
import { Formik, FormikProps } from 'formik';
import { ScreenProps, showNotification, i18n } from '@app/core';
import { navigationService, authService } from '@app/services';
import { styles } from './styles';
import { mapStateToProps } from './map_state_to_props';
import { mapDispatchToProps } from './map_dispatch_to_props';

type Props = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps> & ScreenProps;

interface FormData {
  email: string;
  isEmailRegistered: boolean;
  password: string;
  confirmPassword: string;
}

export const Screen = ({ componentId, language, login }: Props): JSX.Element => {
  const { t } = useTranslation();
  const [isBusy, setIsBusy] = useState<boolean>(false);

  const initialValues: FormData = {
    email: '',
    isEmailRegistered: false,
    password: '',
    confirmPassword: '',
  };
  const fieldNames = {
    email: 'email',
    isEmailRegistered: 'isEmailRegistered',
    password: 'password',
    confirmPassword: 'confirmPassword',
  };
  const validationSchema = Yup.object().shape({
    [fieldNames.email]: Yup.string()
      .email(
        t('error.invalid', {
          field: t('emailRegisterScreen.email'),
        }),
      )
      .required(
        t('error.required', {
          field: t('emailRegisterScreen.email'),
        }),
      ),
    [fieldNames.isEmailRegistered]: Yup.boolean().oneOf(
      [false],
      t('emailRegisterScreen.emailHasBeenAlreadyRegistered'),
    ),
    [fieldNames.password]: Yup.string()
      .matches(
        /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,20}$/,
        t('error.invalid', {
          field: t('emailRegisterScreen.password'),
        }),
      )
      .required(
        t('error.required', {
          field: t('emailRegisterScreen.password'),
        }),
      ),
    [fieldNames.confirmPassword]: Yup.string()
      .oneOf([Yup.ref('password')], t('emailRegisterScreen.confirmPasswordDoesNotMatch'))
      .required(
        t('error.required', {
          field: t('emailRegisterScreen.confirmPassword'),
        }),
      ),
  });

  let formikProps: FormikProps<FormData>;

  const validateUniqueEmail = async (email: string): Promise<void> => {
    if (formikProps.errors.email) {
      return;
    }
    const isEmailRegistered = await authService.isEmailRegistered(email);
    const message = isEmailRegistered ? t('emailRegisterScreen.emailHasBeenAlreadyRegistered') : undefined;
    formikProps.setFieldValue('isEmailRegistered', isEmailRegistered);
    formikProps.setFieldError('isEmailRegistered', (message as unknown) as string);
  };

  const debounceValidateUniqueEmail = _.debounce(validateUniqueEmail, 500);

  const handleChangeEmail = (email: string): void => {
    formikProps.handleChange(fieldNames.email)(email);
    debounceValidateUniqueEmail(email);
  };

  const onSubmit = async (values: FormData): Promise<void> => {
    try {
      setIsBusy(true);
      await validateUniqueEmail(values.email);
      if (!formikProps.isValid) {
        return;
      }
      const user = await authService.createUserWithEmailAndPassword(values.email, values.password, language);
      login(user);
      showNotification({
        type: 'SUCCESS',
        message: t('emailRegisterScreen.accountHasBeenCreated'),
      });
      navigationService.setRootEmailVerification();
    } finally {
      setIsBusy(false);
    }
  };

  return (
    <Container showHeader showBackButton componentId={componentId} headerTitle={t('emailRegisterScreen.register')}>
      <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={onSubmit}>
        {(props) => {
          formikProps = props;
          const { values, touched, errors, handleChange, handleBlur, handleSubmit } = props;
          return (
            <>
              <Field
                label={t('emailRegisterScreen.email')}
                value={values.email}
                onChangeText={handleChangeEmail}
                onBlur={handleBlur(fieldNames.email)}
                error={(touched.email && !!errors.email) || (touched.email && !!errors.isEmailRegistered)}
                success={touched.email && !errors.email && !errors.isEmailRegistered}
                errorMessage={touched.email && !!errors.email ? errors.email : errors.isEmailRegistered}
              />
              <Field
                label={t('emailRegisterScreen.password')}
                value={values.password}
                onChangeText={handleChange(fieldNames.password)}
                onBlur={handleBlur(fieldNames.password)}
                error={touched.password && !!errors.password}
                success={touched.password && !errors.password}
                errorMessage={errors.password}
                secureTextEntry
                hasTooltip
                tooltipText={t('common.passwordInfo')}
                tooltipHeight={language === i18n.LANGUAGE_EN ? 140 : 100}
              />
              <Field
                label={t('emailRegisterScreen.confirmPassword')}
                value={values.confirmPassword}
                onChangeText={handleChange(fieldNames.confirmPassword)}
                onBlur={handleBlur(fieldNames.confirmPassword)}
                error={touched.confirmPassword && !!errors.confirmPassword}
                success={touched.confirmPassword && !errors.confirmPassword}
                errorMessage={errors.confirmPassword}
                secureTextEntry
              />
              <Button onPress={handleSubmit} disabled={isBusy} style={styles.button}>
                <Text white>{t('emailRegisterScreen.register')}</Text>
              </Button>
            </>
          );
        }}
      </Formik>
    </Container>
  );
};
