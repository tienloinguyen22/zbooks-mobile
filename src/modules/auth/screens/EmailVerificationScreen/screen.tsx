import React, { useState } from 'react';
import { Linking, Platform } from 'react-native';
import { useTranslation } from 'react-i18next';
import produce from 'immer';
import { ScreenProps, showNotification } from '@app/core';
import { useEffectOnce } from '@app/hooks';
import { Button, Text, Container } from '@app/components';
import { authService, navigationService } from '@app/services';
import { mapStateToProps } from './map_state_to_props';
import { mapDispatchToProps } from './map_dispatch_to_props';
import { styles } from './styles';

type Props = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps> & ScreenProps;
interface ResendVerificationEmailStatus {
  isVerificationEmailSent: boolean;
  waitingTimeToResend: number;
}

export const Screen = ({ componentId, markEmailVerified, currentUser, logout }: Props): JSX.Element => {
  let resendInterval: NodeJS.Timeout | undefined;
  const { t } = useTranslation();
  const [isBusy, setIsBusy] = useState<boolean>(false);
  const [resendVerificationEmailStatus, setResendVerificationEmailStatus] = useState<ResendVerificationEmailStatus>({
    isVerificationEmailSent: false,
    waitingTimeToResend: 0,
  });

  useEffectOnce(() => {
    return () => {
      if (resendInterval) {
        clearInterval(resendInterval);
      }
    };
  });

  const checkStatus = async (): Promise<void> => {
    try {
      setIsBusy(true);
      if (await authService.isEmailVerified()) {
        markEmailVerified();
        navigationService.setRootHome();
      } else {
        showNotification({
          type: 'WARNING',
          message: t('emailVerificationScreen.emailNotVerified'),
        });
      }
    } finally {
      setIsBusy(false);
    }
  };

  const resendVerificationEmail = async (): Promise<void> => {
    try {
      setIsBusy(true);
      await authService.resendVerificationEmail();
      showNotification({
        type: 'SUCCESS',
        message: t('emailVerificationScreen.resendEmailNotification'),
      });
      setResendVerificationEmailStatus({
        isVerificationEmailSent: true,
        waitingTimeToResend: 30,
      });
      resendInterval = setInterval(() => {
        setResendVerificationEmailStatus(
          produce((draftState: ResendVerificationEmailStatus) => {
            if (!draftState.isVerificationEmailSent) {
              draftState.isVerificationEmailSent = true;
            }
            if (draftState.waitingTimeToResend > 1) {
              draftState.waitingTimeToResend -= 1;
            } else if (draftState.waitingTimeToResend === 1) {
              draftState.isVerificationEmailSent = false;
              draftState.waitingTimeToResend = 0;
            }
          }),
        );
      }, 1000);
    } finally {
      setIsBusy(false);
    }
  };

  const openMailbox = async (): Promise<void> => {
    try {
      setIsBusy(true);
      if (Platform.OS === `ios`) {
        await Linking.openURL(`message:`);
      }
    } finally {
      setIsBusy(false);
    }
  };

  const performLogout = async (): Promise<void> => {
    await authService.logout();
    logout();
    navigationService.setRootLogin();
  };

  const waitForResend =
    resendVerificationEmailStatus.isVerificationEmailSent && resendVerificationEmailStatus.waitingTimeToResend > 0;
  return (
    <Container showHeader componentId={componentId} headerTitle={t('emailVerificationScreen.verifyEmail')} center>
      <Text style={styles.notice}>
        {t('emailVerificationScreen.notice', {
          email: currentUser.email,
        })}
      </Text>
      <Button onPress={checkStatus} disabled={isBusy} style={styles.button}>
        <Text white>{t('emailVerificationScreen.check')}</Text>
      </Button>
      <Button onPress={resendVerificationEmail} disabled={isBusy || waitForResend} style={styles.button}>
        <Text white>
          {t('emailVerificationScreen.resendVerification')}
          {waitForResend ? ` (${resendVerificationEmailStatus.waitingTimeToResend})` : ''}
        </Text>
      </Button>
      {Platform.OS === 'ios' && (
        <Button onPress={openMailbox} disabled={isBusy} style={styles.button}>
          <Text white>{t('emailVerificationScreen.openMailbox')}</Text>
        </Button>
      )}
      <Button onPress={performLogout} disabled={isBusy} style={styles.button}>
        <Text white>{t('emailVerificationScreen.useAnotherAccount')}</Text>
      </Button>
    </Container>
  );
};
