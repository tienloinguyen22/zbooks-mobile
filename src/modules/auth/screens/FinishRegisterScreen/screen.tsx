import React from 'react';
import { View } from 'react-native';
import { Text } from '@app/components';
import { ScreenProps } from '@app/core';

type Props = ScreenProps;

export const Screen = (_props: Props): JSX.Element => {
  return (
    <View>
      <Text>Finish register</Text>
    </View>
  );
};
