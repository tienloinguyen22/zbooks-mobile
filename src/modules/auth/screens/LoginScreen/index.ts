import { Screen } from './screen';

export const LoginScreen = Screen;
