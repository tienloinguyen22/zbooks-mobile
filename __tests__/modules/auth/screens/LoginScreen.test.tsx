import { render, fireEvent } from '@testing-library/react-native';
import { Screen } from '@app/modules/auth/screens/LoginScreen/screen';
import React from 'react';
import { navigationService, authService, LoginResult } from '@app/services';

jest.mock('@app/store', () => {
  return {
    connect: jest.fn().mockReturnValue(() => jest.fn()),
    store: {
      getState: () => ({
        settings: {
          primaryColorCode: 'red',
          theme: 'theme',
        },
      }),
    },
  };
});

jest.mock('@app/modules/auth/screens/LoginScreen/map_state_to_props', () => {
  return {
    mapStateToProps: jest.fn().mockReturnValue(() => ({
      shouldShownUpdateWarning: true,
    })),
  };
});

jest.mock('@app/modules/auth/screens/LoginScreen/map_dispatch_to_props', () => {
  return {
    login: jest.fn().mockReturnValue(() => ({})),
    updateShownUpdateWarning: jest.fn().mockReturnValue(() => ({})),
  };
});

describe('modules/auth/screen/LoginScreen', () => {
  it('renders successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();

    const { baseElement } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('logins facebook successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    authService.loginFacebook = jest.fn().mockImplementation(
      async (): Promise<LoginResult> => {
        return {
          isSuccessful: true,
          user: {
            id: 'id',
            isLoggedIn: true,
            loginType: 'FACEBOOK',
          },
        };
      },
    );
    navigationService.setRootHome = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginFacebook = getByText('loginScreen.loginWith Facebook');
    fireEvent.press(buttonLoginFacebook);

    expect(authService.loginFacebook).toBeCalledTimes(1);
    // expect(navigationService.setRootHome).toBeCalledTimes(1);
  });

  it('logins facebook with error message', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    authService.loginFacebook = jest.fn().mockImplementation(
      async (): Promise<LoginResult> => {
        return {
          isSuccessful: false,
          isCancelled: false,
          errorMessage: 'message',
        };
      },
    );
    navigationService.setRootHome = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginFacebook = getByText('loginScreen.loginWith Facebook');
    fireEvent.press(buttonLoginFacebook);

    expect(authService.loginFacebook).toBeCalledTimes(1);
    expect(navigationService.setRootHome).toBeCalledTimes(0);
  });

  it('logins facebook without error message', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    authService.loginFacebook = jest.fn().mockImplementation(
      async (): Promise<LoginResult> => {
        return {
          isSuccessful: false,
          isCancelled: false,
          errorMessage: '',
        };
      },
    );
    navigationService.setRootHome = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginFacebook = getByText('loginScreen.loginWith Facebook');
    fireEvent.press(buttonLoginFacebook);

    expect(authService.loginFacebook).toBeCalledTimes(1);
    expect(navigationService.setRootHome).toBeCalledTimes(0);
  });

  it('logins facebook with throw error', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    authService.loginFacebook = jest.fn().mockImplementation(() => {
      throw new Error('Login error');
    });
    navigationService.setRootHome = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={false}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginFacebook = getByText('loginScreen.loginWith Facebook');
    fireEvent.press(buttonLoginFacebook);

    expect(authService.loginFacebook).toBeCalledTimes(1);
    expect(navigationService.setRootHome).toBeCalledTimes(0);
  });

  it('presses login google successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    authService.loginGoogle = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginGoogle = getByText('loginScreen.loginWith Google');
    fireEvent.press(buttonLoginGoogle);

    expect(authService.loginGoogle).toBeCalledTimes(1);
  });

  it('presses login with email successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    navigationService.navigateTo = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginEmail = getByText('loginScreen.loginWithEmail');
    fireEvent.press(buttonLoginEmail);

    expect(navigationService.navigateTo).toBeCalledTimes(1);
  });

  it('presses login with phoneNo successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    navigationService.navigateTo = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonLoginWithPhoneNo = getByText('loginScreen.loginWithPhoneNo');
    fireEvent.press(buttonLoginWithPhoneNo);

    expect(navigationService.navigateTo).toBeCalledTimes(1);
  });

  it('presses registerEmail successfully', async () => {
    const login = jest.fn();
    const updateShownUpdateWarning = jest.fn();
    navigationService.navigateTo = jest.fn();

    const { getByText } = render(
      <Screen
        login={login}
        componentId={''}
        shouldShownUpdateWarning={true}
        updateShownUpdateWarning={updateShownUpdateWarning}
      />,
    );
    const buttonRegisterEmail = getByText('loginScreen.registerByEmail');
    fireEvent.press(buttonRegisterEmail);

    expect(navigationService.navigateTo).toBeCalledTimes(1);
  });
});
