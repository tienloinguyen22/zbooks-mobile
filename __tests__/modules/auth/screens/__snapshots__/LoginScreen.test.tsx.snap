// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/auth/screen/LoginScreen renders successfully 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Array [
          Object {
            "flex": 0,
          },
          Object {
            "backgroundColor": "#d32f2f",
          },
        ]
      }
    />
    <SafeAreaView
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#F7F9FC",
          },
        ]
      }
    >
      <View
        center={true}
        centerVertical={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F7F9FC",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/images/app_icon.png",
            }
          }
          style={
            Object {
              "height": 150,
              "margin": 20,
              "width": 150,
            }
          }
        />
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            rounded={true}
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d32f2f",
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "center",
                  "width": 200,
                },
                Object {
                  "alignSelf": "center",
                  "marginBottom": 20,
                  "width": 300,
                },
                Object {
                  "backgroundColor": "#4267B2",
                },
                Object {
                  "borderRadius": 20,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 25,
                  },
                  Object {
                    "fontWeight": "bold",
                  },
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textTransform": "none",
                }
              }
              white={true}
            >
              loginScreen.loginWith
               Facebook
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            rounded={true}
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d32f2f",
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "center",
                  "width": 200,
                },
                Object {
                  "alignSelf": "center",
                  "marginBottom": 20,
                  "width": 300,
                },
                Object {
                  "backgroundColor": "#EA4334",
                },
                Object {
                  "borderRadius": 20,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 25,
                  },
                  Object {
                    "fontWeight": "bold",
                  },
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textTransform": "none",
                }
              }
              white={true}
            >
              loginScreen.loginWith
               Google
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            rounded={true}
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d32f2f",
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "center",
                  "width": 200,
                },
                Object {
                  "alignSelf": "center",
                  "marginBottom": 20,
                  "width": 300,
                },
                Object {
                  "borderRadius": 20,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textTransform": "none",
                }
              }
              white={true}
            >
              loginScreen.loginWithEmail
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            rounded={true}
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d32f2f",
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "center",
                  "width": 200,
                },
                Object {
                  "alignSelf": "center",
                  "marginBottom": 20,
                  "width": 300,
                },
                Object {
                  "borderRadius": 20,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textTransform": "none",
                }
              }
              white={true}
            >
              loginScreen.loginWithPhoneNo
            </Text>
          </View>
        </TouchableOpacity>
        <Text
          style={
            Object {
              "color": "#222C44",
              "marginBottom": 20,
              "textTransform": "none",
            }
          }
        >
          loginScreen.notHaveAccount
        </Text>
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <View
            rounded={true}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#d32f2f",
                "borderRadius": 20,
                "flexDirection": "row",
                "height": 45,
                "justifyContent": "center",
                "marginBottom": 20,
                "width": 300,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                  "textTransform": "none",
                }
              }
              white={true}
            >
              loginScreen.registerByEmail
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  </View>
</View>
`;
